### PR TITLE
Backward support for unbind() with NJT

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128032

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -2852,7 +2852,7 @@
       self: unbind_backward(grads, dim)
       result: auto_linear
     AutogradNestedTensor:
-      self: unbind_backward_nested(grads, at::native::get_nested_tensor_impl(self)->get_nested_sizes(), dim, self.options())
+      self: "self.layout() == c10::kJagged ? unbind_backward_nested_jagged(grads, self, dim) : unbind_backward_nested(grads, at::native::get_nested_tensor_impl(self)->get_nested_sizes(), dim, self.options())"
```
Comment (soulitzer) [inline tools/autograd/derivatives.yaml:2855]: FWIW this means that the NST case also saves self now

Comment (jbschlosser) [reply tools/autograd/derivatives.yaml:2855]: that is sad yes.. it's not ideal

Comment (jbschlosser) [reply tools/autograd/derivatives.yaml:2850]: wait, it's already saving self due to `at::native::get_nested_tensor_impl(self)->get_nested_sizes()`

**Comment by jbschlosser** [comment]
From offline discussion with @soulitzer, the ideal long-term solution is to have a single derivative formula across NJT / NST that saves as little as possible for backward: `unbind_backward_nested(grads, self.sym_sizes(), dim, self.options())`. This requires:
* Nested int -> {offsets / min seq len / max seq len} accessible from C++ (so metadata can be propagated to the returned grad_input)
* Nested int support for NST shapes (so `nst.sym_sizes()` works)

I played around with some workarounds:
* Store a meta-fied version of `self` to minimize memory consumption in what's saved for backward
    * This doesn't fully work because we need the real `offsets`
* Implement `unbind()` via torch_function instead of torch_dispatch and redispatch to `split()` / `narrow()`
    * This pushes the problem down to the other view ops; now they need to support operating on the ragged dim with correct derivative formulas. Going to continue down this path for a bit

**Review by soulitzer — APPROVED** [review]

**Comment by jbschlosser** [comment]
Follow-ups (for future PRs):
* Stop saving `self` for backward for NJT's `values()`
* Stop saving `self` for backward for NST's metadata; this is useful if we want to use the same formula for both NJT / NST
* Revisit torch_function version of `unbind()` using `values()` + some other view ops
    * Benchmark dispatch overhead

**Pull request closed without merging**